addons/mrp/models/mrp_production.py

```diff
@@ -1715,28 +1711,21 @@ def _get_quantity_to_backorder(self):
 
     def _check_sn_uniqueness(self):
         """ Alert the user if the serial number as already been consumed/produced """
-        if self.product_tracking == 'serial' and self.lot_producing_id:
-            sml = self.env['stock.move.line'].search_count([
-                ('lot_id', '=', self.lot_producing_id.id),
-                ('location_id.usage', '=', 'production'),
-                ('qty_done', '=', 1),
-                ('state', '=', 'done')
-            ])
-            if sml:
-                raise UserError(_('This serial number for product %s has already been produced', self.product_id.name))
-
-        for move in self.move_finished_ids:
-            if move.has_tracking != 'serial' or move.product_id == self.product_id:
-                continue
-            for move_line in move.move_line_ids:
+        for move in self.move_finished_ids.filtered(lambda m: m.has_tracking == 'serial'):
+            to_check = move.move_line_ids.filtered(lambda l: l.lot_id).mapped(lambda l: (l.product_id, l.lot_id, l))
+            if self.product_tracking == 'serial' and self.product_id not in move.move_line_ids.product_id and self.lot_producing_id:
```

Comment:
It's probably best if you leave this a separate check like it originally was because now you check the production SN for every byproduct move you have (i.e. 1 + num_of_byproduct_moves checks). Then you can also use the private method that you create for the onchange.